### PR TITLE
Fix: Prefix function declarations in TS namespaces (fixes #78)

### LINKED ISSUE #78: Escope null body failure
**What version of TypeScript are you using?**

 1.8.10

**What version of `typescript-eslint-parser` are you using?**

 0.2.0

**What code were you trying to parse?** 

d3.d.ts from 
https://raw.githubusercontent.com/DefinitelyTyped/DefinitelyTyped/master/d3/d3.d.ts

**What did you expect to happen?**

Correct behavior

**What happened?**

```
node_modules\eslint\node_modules\escope\lib\scope.js:113
        for (i = 0, iz = body.body.length; i < iz; ++i) {
                             ^

TypeError: Cannot read property 'body' of null
    at isStrictScope (node_modules\eslint\node_modules\escope\lib\scope.js:113:30)
    at FunctionScope.Scope (node_modules\eslint\node_modules\escope\lib\scope.js:249:25)
    at new FunctionScope (node_modules\eslint\node_modules\escope\lib\scope.js:689:92)
    at ScopeManager.__nestFunctionScope (node_modules\eslint\node_modules\escope\lib\scope-manager.js:241:37)
    at Referencer.visitFunction (node_modules\eslint\node_modules\escope\lib\referencer.js:236:31)
    at Referencer.FunctionDeclaration (node_modules\eslint\node_modules\escope\lib\referencer.js:564:18)
    at Referencer.Visitor.visit (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:122:34)
    at Referencer.visitExportDeclaration (node_modules\eslint\node_modules\escope\lib\referencer.js:603:22)
    at Referencer.ExportNamedDeclaration (node_modules\eslint\node_modules\escope\lib\referencer.js:617:18)
    at Referencer.Visitor.visit (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:122:34)
    at Referencer.Visitor.visitChildren (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:101:38)
    at Referencer.Visitor.visit (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:125:14)
    at Referencer.Visitor.visitChildren (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:106:26)
    at Referencer.Visitor.visit (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:125:14)
    at Referencer.Visitor.visitChildren (node_modules\eslint\node_modules\escope\node_modules\esrecurse\esrecurse.js:101:38)
    at Referencer.Program (node_modules\eslint\node_modules\escope\lib\referencer.js:419:18)

```

### PR DESCRIPTION
**Pull request by JamesHenry**

### CODE REVIEW

File: lib/ast-converter.js [hunk 1]
```diff
@@ -215,8 +215,18 @@ function fixExports(node, result, ast) {
         result.range[0] = varToken.getStart();
         result.loc = getLocFor(result.range[0], result.range[1], ast);
 
+        var declarationType = declarationIsDefault ? "ExportDefaultDeclaration" : "ExportNamedDeclaration";
+
+        /**
+         * Prefix exports from TypeScript namespaces with "TS" to distinguish
+         * them from ES2015 exports
+         */
+        if (node.parent && node.parent.kind === SyntaxKind.ModuleBlock) {
```
Comment (nzakas) [inline lib/ast-converter.js]: I wonder if we should be more specific with these names. 

TSNamespaceExportDeclaration?

File: lib/ast-converter.js [hunk 2]
```diff
@@ -776,6 +786,13 @@ module.exports = function(ast, extra) {
                     }
                 }
 
+                /**
+                 * Prefix FunctionDeclarations within TypeScript namespaces with "TS"
+                 */
+                if (node.parent && node.parent.kind === SyntaxKind.ModuleBlock) {
+                    functionDeclarationType = "TS" + functionDeclarationType;
```
Comment (nzakas) [inline lib/ast-converter.js]: TSNamespaceFunctionDeclaration?

Is there any other location where this name might be used?

Comment (JamesHenry) [inline lib/ast-converter.js]: I'm happy with `TSNamespaceFunctionDeclaration`. Which name are you referring to?

File: lib/ast-converter.js (hunk 1 above)
Comment (JamesHenry) [inline lib/ast-converter.js]: Fine by me

**Comment by JamesHenry** [comment]
PTAL

File: lib/ast-converter.js (hunk 2 above)
Comment (nzakas) [inline lib/ast-converter.js]: `TSNamespaceFunctionDeclaration`. I'm just wondering if this same syntactic form can be used in other locations or not. (Basically, is "Namespace" too specific for this node or not?)

**Comment by JamesHenry** [comment]
(Just moving this into the main comment flow, as the comments on the outdated diff are not shown by default)

From @nzakas:
> TSNamespaceFunctionDeclaration. I'm just wondering if this same syntactic form can be used in other locations or not. (Basically, is "Namespace" too specific for this node or not?)

Me: After reviewing the spec again, I do not believe so, and I think this is PR should be good to go

**Pull request merged**